app/webpacker/components/Live/Admin/DoubleCheck/index.jsx

```diff
@@ -0,0 +1,151 @@
+import React, {
+  useCallback, useEffect, useMemo, useState,
+} from 'react';
+import {
+  Button, Card, Form,
+  Grid, Header,
+  Segment,
+} from 'semantic-ui-react';
+import { useMutation, useQuery } from '@tanstack/react-query';
+import { events } from '../../../../lib/wca-data.js.erb';
+import WCAQueryClientProvider from '../../../../lib/providers/WCAQueryClientProvider';
+import AttemptResultField from '../../../EditResult/WCALive/AttemptResultField/AttemptResultField';
+import updateRoundResults from '../../api/updateRoundResults';
+import getRoundResults from '../../api/getRoundResults';
+import Loading from '../../../Requests/Loading';
+
+export default function Wrapper({
+  competitionId, competitors, round,
+}) {
+  return (
+    <WCAQueryClientProvider>
+      <Competitors
+        eventId={round.event.id}
+        competitionId={competitionId}
+        competitors={competitors}
+        round={round}
+      />
+    </WCAQueryClientProvider>
+  );
+}
+
+function zeroedArrayOfSize(size) {
+  return Array(size).fill(0);
+}
+
+function Competitors({
+  competitionId, competitors, round, eventId,
+}) {
+  const event = events.byId[eventId];
+  const solveCount = event.recommendedFormat().expectedSolveCount;
+  const [currentIndex, setCurrentIndex] = useState(0);
+  const [registrationId, setRegistrationId] = useState(competitors[0].id);
+  const [attempts, setAttempts] = useState(zeroedArrayOfSize(solveCount));
+
+  const { data: results, isLoading } = useQuery({
+    queryKey: [round.id, 'results'],
+    queryFn: () => getRoundResults(round.id, competitionId),
+  });
+
+  const {
+    mutate: mutateUpdate, isPending: isPendingUpdate,
+  } = useMutation({
+    mutationFn: updateRoundResults,
+  });
+
+  const handleSubmit = async () => {
+    mutateUpdate({
+      roundId: round.id, registrationId, competitionId, attempts,
+    });
+  };
+
+  const handleAttemptChange = (index, value) => {
+    const newAttempts = [...attempts];
+    newAttempts[index] = value;
+    setAttempts(newAttempts);
+  };
+
+  const handleRegistrationIdChange = useCallback((_, { value }) => {
+    setRegistrationId(value);
+    const alreadyEnteredResults = results.find((r) => r.registration_id === value);
+    setAttempts(alreadyEnteredResults.attempts.map((a) => a.result));
+  }, [results]);
+
+  const currentCompetitor = useMemo(() => {
+    if (results) {
+      return competitors.find((r) => r.id === results[currentIndex].registration_id);
+    }
+    return {};
+  }, [competitors, currentIndex, results]);
+
+  useEffect(() => {
+    if (results) {
+      setAttempts(results[currentIndex].attempts.map((a) => a.result));
+    }
+  }, [currentIndex, results]);
```

Comment:
This looks like you can achieve the same thing without `useEffect`.

Reply:
Isn't this essentially a `useMemo` depending on `results`?

Reply:
But `currentIndex` is also just a React state. If you put that into the memo dependencies, it will automatically update. I therefore insist this `useEffect` is unnecessary.

Reply:
I don't get it: 
```
  useEffect(() => {
    if (results) {
      setAttempts(results[currentIndex].attempts.map((a) => a.result));
    }
  }, [currentIndex, results]);
```
and 
```
 useMemo(() => {
    if (results) {
      setAttempts(results[currentIndex].attempts.map((a) => a.result));
    }
  }, [currentIndex, results]);
```
are functionally identical. and I don't think a useMemo is supposed to trigger a side effect like that

Reply:
and you can't just do 
```
const attempts = useMemo(() => {
    if (results) {
      results[currentIndex].attempts.map((a) => a.result);
    }
  }, [currentIndex, results]);
```
Because attempts needs to be a state so it can be manipulated from the form

Reply:
Is the query ever re-fetched? i.e. Does `results` ever change? If not, then you indeed do need a side effect to set `attempts` when the data first loads, but you could explicitly update `attempts` when the "<" or ">" button is clicked instead of it being a side effect of the index changing. That would minimize the use of effects (even though it's the same number of `useEffects`, that effect would only be happening once).